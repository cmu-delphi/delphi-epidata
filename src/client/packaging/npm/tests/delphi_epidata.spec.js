/**
 * @global {jest}
 */
const fetchM = require("cross-fetch");
const { Epidata, EpidataAsync } = require("..");

jest.mock("cross-fetch");

beforeEach(() => {
  fetchM.fetch.mockClear();
});

function mockReturn(res) {
  fetchM.fetch.mockResolvedValue({
    json: () => Promise.resolve(res),
  });
}

describe("fluview", () => {
  describe("basic", () => {
    const res = { message: "m", result: 1, epidata: [] };
    it("async", async () => {
      mockReturn(res);
      const r = await EpidataAsync.fluview("a", 4);
      expect(r).toEqual(res);
      expect(fetchM.fetch).toMatchSnapshot();
    });
    it("sync", (done) => {
      mockReturn(res);
      Epidata.fluview(
        (result, message, epidata) => {
          expect({ result, message, epidata }).toEqual(res);
          expect(fetchM.fetch).toMatchSnapshot();
          done();
        },
        "a",
        4
      );
    });
  });
  it("wrong params", () => {
    expect(() => EpidataAsync.fluview()).toThrow(
      "`regions`,`epiweeks` are all required"
    );
    expect(() => EpidataAsync.fluview("a")).toThrow(
      "`regions`,`epiweeks` are all required"
    );
    expect(() => EpidataAsync.fluview(null)).toThrow(
      "`regions`,`epiweeks` are all required"
    );
    expect(() => EpidataAsync.fluview("a", 4, 4, 3)).toThrow(
      "`issues` and `lag` are mutually exclusive"
    );
  });
});

describe("alternative API endpoints", () => {
  const res = { message: "m", result: 1, epidata: [] };
  it("php", async () => {
    mockReturn(res);
    const api = EpidataAsync.withURL("http://test.com/api.php");
    const r = await api.fluview("a", 4);
    expect(r).toEqual(res);
    expect(fetchM.fetch).toHaveBeenCalledTimes(1);
    expect(fetchM.fetch.mock.calls[0][0].toString()).toEqual(
      "http://test.com/api.php?endpoint=fluview&regions=a&epiweeks=4"
    );
    expect(fetchM.fetch).toMatchSnapshot();
  });
  it("new", async () => {
    mockReturn(res);
    const api = EpidataAsync.withURL("http://test.com/");
    const r = await api.fluview("a", 4);
    expect(r).toEqual(res);
    expect(fetchM.fetch).toHaveBeenCalledTimes(1);
    expect(fetchM.fetch.mock.calls[0][0].toString()).toEqual(
      "http://test.com/fluview/?regions=a&epiweeks=4"
    );
    expect(fetchM.fetch).toMatchSnapshot();
  });
});
