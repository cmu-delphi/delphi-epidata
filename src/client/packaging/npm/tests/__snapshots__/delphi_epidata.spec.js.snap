// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`alternative API endpoints new 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "http://test.com/fluview/?regions=a&epiweeks=4",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`alternative API endpoints php 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "http://test.com/api.php?endpoint=fluview&regions=a&epiweeks=4",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`fluview basic async 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://api.delphi.cmu.edu/epidata/fluview/?regions=a&epiweeks=4",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`fluview basic sync 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://api.delphi.cmu.edu/epidata/fluview/?regions=a&epiweeks=4",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;
